### ISSUE TITLE
Slow compilation when using many .ThenInclude()'s

### ISSUE THREAD

**Issue by dharmaturtle** [aspnet/EntityFrameworkCore#17455]
In EFCore3 Preview 7 onwards, there is a significant slow down when I have a query with many `.ThenInclude()`s, e.g.

```
new WhyYouSoSlowDb(options)
        .Blog
        .Include(x => x.Posts)
          .ThenInclude(x => x.PostInstances)
          .ThenInclude(x => x.Comments)
          .ThenInclude(x => x.AcquiredComments)
          .ThenInclude(x => x.Tag_AcquiredComments)
          .ThenInclude(x => x.Tag)
        .FirstOrDefault();
```
The above takes 20 seconds to run on an empty database. On EFCore3 Preview 5, it takes 1.5 seconds to run. (Preview 6 throws an exception.)

### Steps to reproduce

Repo available [here](https://github.com/dharmaturtle/EFSpeedRegression). Note that some amount of complexity is required to reproduce this issue - if you delete the 5 unreferenced properties from `AcquiredCommentEntity`, it will take 10 seconds to run.

### Further technical details
EF Core version: EFCore3 Preview 7 and 8
Database Provider: Microsoft.EntityFrameworkCore.SqlServer
Operating system: Windows 10
IDE: Visual Studio 2019 Preview

**Comment by Enngage** [comment]
I started seeing this as well when upgrading from preview 5 to preview 8. Queries that took 100ms now take up to 2 minutes - that includes repeated tests. I'm not sure if caching helps in this case since I wouldn't really want to wait 2 minutes to get relatively simple data (though query has like 10 nested includes). 

Based on this discussion, is this something that is going to be addressed in next preview / final release or should we rewrite queries?

**Comment by AndriySvyryd** [comment]
Related to https://github.com/aspnet/EntityFrameworkCore/issues/12632

**Comment by Enngage** [comment]
@AndriySvyryd I'm not sure if that is related as in #12632 the issue seems to be slow performing SQL queries and SQL timeouts, whereas in this case it is the query compilation that is slow. Actual SQL is quite fast (0.5s) - though still slower then previous implementation in `preview5`.

**Comment by roji** [comment]
I did some further investigation; my SelectExpression optimizations above were OK but didn't address the root of the problem at all.

Profiling this code shows that SelectExpression.VisitChildren (as well as other visit methods) gets called an incredible number of times (upwards of 2 million times). The reason for this, is that all the projections in this query contain ColumnExpressions whose table is a reference to other SelectExpression. This causes repeated visitation of the same SelectExpression over and over, with the number of visitations increasing exponentially as more includes are added.

For example, here's the top-most select:
```sql
SELECT [t].[Id], [t].[Name], [t].[UserId], [t6].[Id], [t6].[BlogId], [t6].[Description], [t6].[UserId], [t6].[Id0], [t6].[Created], [t6].[Hash], [t6].[IsDeleted], [t6].[Modified], [t6].[PostId], [t6].[Id00], [t6].[Index], [t6].[PostInstanceId], [t6].[TemplateId], [t6].[UserId0], [t6].[CommentId], [t6].[CommentState], [t6].[Created0], [t6].[IsDeleted0], [t6].[Meta], [t6].[Meta2], [t6].[MetaId], [t6].[TagId], [t6].[UserId00], [t6].[CommentId0], [t6].[Id000], [t6].[Name], [t6].[UserId000]
```

This SelectExpressions contains 28 projections to columns on `[t6]`, which itself is a SelectExpression with a large number of projections over other SelectExpressions, and so forth recursively.

One immediate question this raises, is if we can simply replace the joins on subqueries with joins  on the tables directly:

```sql
SELECT ...
FROM (
    SELECT TOP(1) [b].[Id], [b].[Name], [b].[UserId]
    FROM [Blog] AS [b]
) AS [t]
LEFT JOIN [Post] AS [t6] ON [t].[Id] = [t6].[BlogId]
LEFT JOIN [PostInstance] AS [t5] ON [p].[Id] = [t5].[PostId]
LEFT JOIN [Comment] AS [t4] ON [p0].[Id] = [t4].[PostInstanceId]
LEFT JOIN [AcquiredComment] AS [t3] ON [c].[Id] = [t3].[CommentId]
... and so on
```

This would likely also help runtime (or at least produce better SQL).

But the root problem would still remain for other cases, and to solve it we'd need to avoid re-visiting nodes that have already been visited. I have some thoughts on how we could do that (efficiently), but I wonder if we've already encountered this kind of issue and discussed it.

**Comment by roji** [comment]
One immediate idea for fixing this... It may simply make sense to avoid visiting SelectExpressions (or even TableExpressions) embedded inside ProjectionExpressions - they are simply "references" to something that will get visited anyway via SelectExpression.Tables. In other words, the expression tree is in fact a DAG, and we could simply not traverse projection edges.

As a quick dirty test, I simply replaced ProjectionExpression.VisitChildren with a version that returns `this`, and compilation completes instantly. This is an incorrect fix because other parts of the ProjectionExpression still need to be visited, but it may point in the right direction.

**Comment by Enngage** [comment]
hey @roji, thanks a lot for looking into this! The sample SQL query you posted seems a lot better than what is currently being generated and would indeed perform better (I think). 

This might also address the second issue I encounter initially - the call which evaluated in < 10 ms on SQL with preview5 takes 0.5s on preview8. Thanks again for support :-)

**Comment by smitpatel** [comment]
> But the root problem would still remain for other cases, and to solve it we'd need to avoid re-visiting nodes that have already been visited. I have some thoughts on how we could do that (efficiently), but I wonder if we've already encountered this kind of issue and discussed it.

#17337

**Comment by roji** [comment]
> #17337

@smitpatel I was wondering how that's handled... But just to be sure - even if #17337 is fixed we would still have the issue of visiting the same SelectExpression many times, right? I mean, reconnecting nodes is orthogonal to not re-visiting them?

**Comment by smitpatel** [comment]
> reconnecting nodes is orthogonal to not re-visiting them?

How could you reconnect nodes when they are supposed to be same node? And if it same node then you have only visited it once.

**Comment by roji** [comment]
> And if it same node then you have only visited it once.

Unless I'm mistaken, what's happening now is that SelectExpression.VisitChildren visits all projections, and then visits all tables/subqueries. Even if both these data structures pointed to the same instance, we'd still visit it more than once, no?

**Comment by smitpatel** [comment]
When both of them points to same instance, is the case when SelectExpression did not re-generate. Sure, it is not that bug and it would still cause performance issue. But the only way you can keep same instance after re-generation is not visit the same SelectExpression multiple times, (no matter where it appears so it can expand more than projection too). And that solves the perf issue already. #17337 already fixes perf issue and I don't see a reason to track it separately.

**Comment by roji** [comment]
Makes sense, thanks. I'll a note on the perf aspects in #17337, and we can close this as a dup of #17337?

**Comment by roji** [comment]
One more unrelated issue is the possibility I mentioned above of generating joins on tables directly instead of joins on sub-queries - do we have an issue tracking that already?

**Comment by smitpatel** [comment]
> generating joins on tables directly instead of joins on sub-queries

I believe cost vs value does not meet bar to do it.

**Comment by roji** [comment]
>> generating joins on tables directly instead of joins on sub-queries
>
> I believe cost vs value does not meet bar to do it.

Shouldn't we do a perf measurement to determine that? I mean execution, not compilation. Opened #17622 to track.

**Comment by NetTecture** [comment]
Will this be fixed for 3.0 release? Because I run projects that do 20-30 includes and if that is NOT fixed then basically the performance regression makes this update a total no go. I can not move queries from significantly sub 1 second EFCore handling to MINUTES.

Also re SQL: not meeting the bar is run - up to the point people actually have to DEBUG complex expression trees and the SQL looks like a nightmare compared to the clean SQL that was usus earlier.

**Comment by dragnilar** [comment]
@NetTecture  I just checked the 3.0 release that came out yesterday, the problem is still there. Code shows the changes made from the commit are not present. I guess we and anyone else who has queries with multiple includes will have to wait. :(

**Comment by roji** [comment]
@dragnilar I've just taken the original repro above and changed it to reference EF Core 3.0.0 RTM, and it completed in under a second (with preview8 it took over 18 seconds). Can you please double-check your configuration and make sure you're using the RTM version?

**Comment by dragnilar** [comment]
@roji The good news is that my environment did have a problem, so after cleaning it up and redownloading the package, I made sure the code changes you made were present in the disassembly. 

The bad news is that even with those changes, the query still performs worse than it does with EF Core 2.6. 

Mine has significantly more Includes than the repro, so that may be one of the reasons why.

**Comment by roji** [comment]
@dragnilar thanks for the confirmation. This issue was about a extremely slow *compilation* of the query, which occurs only once at startup - I assume you're saying that your query *executes* more slowly, which is a different matter.

Note that EF Core 3.0 contains a significant change whereby a single LINQ query is executed via a single SQL query. In particular, this means that we now generate single-query JOINs in cases where we previously generated multiple queries (one per include), and the duplication of data inherent in joins may be slowing down your query.

Can you please open a separate issue with a small code sample, and if possible post the SQL generated from both 2.2.6 and 3.0?

**Comment by roji** [comment]
To clarify my previous comment, it isn't yet certain whether the problem you're seeing is a compilation or execution problem, so a minimal code sample would help us figure out what's going on.

**Comment by dragnilar** [comment]
@roji  Thanks for the clarification. I noticed the compilation is faster (my integration tests and app looked hung before the query actually fired prior to your fix). 

I think you are correct though, I noticed the SQL being emitted by EF with 2.2.6 is being broken up into small queries versus 3.0 where it's one gargantuan query of joins.

**Comment by smitpatel** [comment]
For no data case, EF Core 2.2.6 is always going to be faster because of multiple SQL.
As for runtime, we still does not have catching of SQL command based on parameter nullability which was present in EF Core 2.2.6 so it would be somewhat slower but we plan to fix that in 3.1

**Comment by eliasbobadilla** [comment]
Hi!, I had a performance alert in the manager that I'm developing.
I added this warning: `RelationalEventId.MultipleCollectionIncludeWarning,` in my database context configuration to found where there are performance problems. After several tests I have identified that `.ThenInclude` is the one that triggers the alert, I am using EF Core 5.0.2 

This is my Code:

```
var applications = await _repo.Application.All()
                .Include(x=> x.Toolkits)
                .ThenInclude(x => x.Tools) //when I remove this line I don't have performance alerts
                .ToListAsync();
```
Was anyone able to solve this performance issue?

**Comment by roji** [comment]
@eliasbobadilla take a look at [this docs page](https://docs.microsoft.com/en-us/ef/core/querying/single-split-queries) to understand single vs. split queries. You may want to make that query use split queries, depending on how many Tools there typically are for each Toolkit (and also how many Toolkits for each Application).

**Comment by eliasbobadilla** [comment]
> @eliasbobadilla take a look at [this docs page](https://docs.microsoft.com/en-us/ef/core/querying/single-split-queries) to understand single vs. split queries. You may want to make that query use split queries, depending on how many Tools there typically are for each Toolkit (and also how many Toolkits for each Application).

thanks a lot @roji , I solved my problem and now I know how to better approach queries with multiple join in sql with Ef
